File: 3208. Alternating Groups II.cpp
class Solution {
public:
    int numberOfAlternatingGroups(const vector<int>& colors, int k) {
        int left = 0, right = 0;
        unordered_set<int> badIndexes;
        for (right = 1; right < k; right++) {
            if (colors[right] == colors[right - 1]) badIndexes.insert(right - 1);
        }
        int ans = 0;
        for (left = 0; left < (int)colors.size(); left++) {
            if (badIndexes.empty()) ans++;
            if (badIndexes.count(left)) badIndexes.erase(left);

            int modRight = right % (int)colors.size();
            int modPrevRight = modRight != 0 ? modRight - 1 : (int)colors.size() - 1;
            if (colors[modPrevRight] == colors[modRight]) badIndexes.insert(modRight - 1);
            right++;
        }
        return ans;
    }
};

File: 3208. Alternating Groups II.java
class Solution {
    public int numberOfAlternatingGroups(int[] colors, int k) {
        int left = 0, right = 0;
        HashSet<Integer> badIndexes = new HashSet<>();
        for(right = 1; right < k; right++){
            if(colors[right] == colors[right - 1]) badIndexes.add(right - 1);
        }
        int ans = 0;
        // if(badIndexes.isEmpty()) ans++;
        for(left = 0; left < colors.length; left++){
            if(badIndexes.isEmpty()) ans++;
            if(badIndexes.contains(left)) badIndexes.remove(left);

            int modRight = right % colors.length;
            int modPrevRight = modRight != 0 ? modRight - 1 : colors.length - 1;
            if(colors[modPrevRight] == colors[modRight]) badIndexes.add(modRight - 1);
            right++;
            // for(int indexes : badIndexes) System.out.print(indexes + " ");
            // System.out.println();

        }

        return ans;
    }
}